### PR TITLE
[WIP] [SmolLM3] Add Backbone and CausalLM

### PR DESCRIPTION
**Pull request by DavidLandup0**
## Description of the change
WIP

## Checklist

- [ ] I have added all the necessary unit tests for my change.
- [ ] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [ ] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [ ] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [ ] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [ ] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: keras_hub/src/models/smollm3/smollm3_backbone.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        num_layers,
+        num_attention_heads,
+        num_key_value_heads,
+        attention_bias,
+        attention_dropout,
+        rope_layer_enabled_list,
+        layer_types,
+        mlp_bias,
+        rms_norm_epsilon,
+        layer_norm_epsilon,
```
Comment (mattdangerw) [inline keras_hub/src/models/smollm3/smollm3_backbone.py:107]: Usually there's some of these terms (like the epsilon's and rope theta) that have a consistent value across all the presets we care about, and we give them defaults here. Not super important, just for people that wanted an easier time making a custom small version of the arch or something like that.

File: keras_hub/src/models/smollm3/smollm3_utils.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        return hidden_states
+    hidden_states = ops.expand_dims(hidden_states, axis=2)
+    target_shape = (batch, num_key_value_heads, n_rep, slen, head_dim)
+    hidden_states = ops.broadcast_to(hidden_states, target_shape)
+    return ops.reshape(
+        hidden_states, [batch, num_key_value_heads * n_rep, slen, head_dim]
+    )
+
+
+def eager_attention_forward(
```
Comment (mattdangerw) [inline keras_hub/src/models/smollm3/smollm3_utils.py]: this is a hf/transforms-ism we'd rather not inherit. perfer just keeping the code on the layer directly, fine to use a private helper to decompose it a bit more.

**Review by mattdangerw — COMMENTED** [review]
Just some quick drive by comments, know this is still wip!

File: keras_hub/src/models/smollm3/smollm3_utils.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    module,
+    query,
+    key,
+    value,
+    attention_mask,
+    scaling: float,
+    dropout: float = 0.0,
```
[the remark this answers is not in the document]
Comment (DavidLandup0) [reply keras_hub/src/models/smollm3/smollm3_utils.py]: Addressed.

File: keras_hub/src/models/smollm3/smollm3_utils.py (hunk 2 above)
Comment (DavidLandup0) [reply keras_hub/src/models/smollm3/smollm3_utils.py]: Very late answer - but yep. Just kept this here during the initial prototyping because it made it easier to functionally test each individual component for numerical equivalence. Refactored to a class-based approach aftewards.

File: keras_hub/src/models/smollm3/smollm3_causal_lm.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        # Get position embeddings for the full sequence
+        position_embeddings = self.backbone.rotary_embedding(x)
+
+        for i, transformer_layer in enumerate(self.backbone.transformer_layers):
+            x = transformer_layer(
+                hidden_states=x,
+                position_embeddings=position_embeddings,
+                attention_mask=padding_mask,
+            )
+            x = layer_intercept_fn(x, i)
```
[the remark this answers is not in the document]
Comment (DavidLandup0) [reply keras_hub/src/models/smollm3/smollm3_causal_lm.py]: Addressed.

File: keras_hub/src/models/smollm3/smollm3_tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        self._add_special_token(eos_token, "end_token")
+
+        bos_token = "<|begin_of_text|>"
+        self._add_special_token(bos_token, "bos_token")
+
+        start_think_token = "<think>"
+        self._add_special_token(start_think_token, "start_think_token")
+
+        end_think_token = "</think>"
+        self._add_special_token(end_think_token, "end_think_token")
```
[the remark this answers is not in the document]
Comment (DavidLandup0) [reply keras_hub/src/models/smollm3/smollm3_tokenizer.py:50]: Addressed.

File: keras_hub/src/models/smollm3/smollm3_layers.py [hunk 6]
```diff
@@ -0,0 +1,699 @@
...
+            key, value = _compute_kv_values(hidden_states)
+
+        if self.use_rope:
+            query = self.rotary_embedding(query, start_index=start_index)
+            key = self.rotary_embedding(key, start_index=start_index)
+
+        print('pre', key.shape, value.shape)
+        key = ops.repeat(key, repeats=self.num_key_value_groups, axis=2)
+        value = ops.repeat(value, repeats=self.num_key_value_groups, axis=2)
+        print('post', key.shape, value.shape)
```
[the remark this answers is not in the document]
Comment (DavidLandup0) [reply keras_hub/src/models/smollm3/smollm3_layers.py]: Addressed.

File: keras_hub/src/models/smollm3/smollm3_layers.py [hunk 7]
```diff
@@ -0,0 +1,699 @@
...
+        else:
+            # Fallback if position_ids_shape is None or malformed.
+            # In this case, the batch_size and seq_len are unknown.
+            batch_size = None
+            seq_len = None
+
+        # The output cos and sin have shape (batch_size, seq_len, head_dim)
+        output_shape = (batch_size, seq_len, self.head_dim)
+
+        return [output_shape, output_shape]
```
[the remark this answers is not in the document]
Comment (DavidLandup0) [reply keras_hub/src/models/smollm3/smollm3_layers.py:757]: Addressed.

File: keras_hub/src/models/smollm3/smollm3_causal_lm_preprocessor.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        ]
+    )
+    labels = tf.constant([1, 0])
+    ds = tf.data.Dataset.from_tensor_slices((features, labels))
+    ds = ds.map(preprocessor, num_parallel_calls=tf.data.AUTOTUNE)
+
+    # Map a dataset to preprocess unlabled sentences.
+    ds = tf.data.Dataset.from_tensor_slices(features)
+    ds = ds.map(preprocessor, num_parallel_calls=tf.data.AUTOTUNE)
+    ```
```
[the remark this answers is not in the document]
Comment (DavidLandup0) [reply keras_hub/src/models/smollm3/smollm3_causal_lm_preprocessor.py:77]: Good point - should update.

File: keras_hub/src/models/smollm3/smollm3_layers.py [hunk 9]
```diff
@@ -0,0 +1,699 @@
...
+        num_key_value_heads: int,
+        attention_bias: bool,
+        attention_dropout: float,
+        rope_layer_enabled_list: list[bool],
+        layer_types: list[str],
+        layer_idx: int,
+        intermediate_size: int,
+        mlp_bias: bool,
+        layer_norm_epsilon: float,
+        **kwargs,
```
[the remark this answers is not in the document]
Comment (DavidLandup0) [reply keras_hub/src/models/smollm3/smollm3_layers.py:441]: Addressed.

File: keras_hub/src/models/smollm3/smollm3_utils.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        return hidden_states
+    hidden_states = ops.expand_dims(hidden_states, axis=2)
+    target_shape = (batch, num_key_value_heads, n_rep, slen, head_dim)
+    hidden_states = ops.broadcast_to(hidden_states, target_shape)
+    return ops.reshape(
+        hidden_states, [batch, num_key_value_heads * n_rep, slen, head_dim]
+    )
+
+
+def rope_init(rope_theta: float, partial_rotary_factor: float, head_dim: int):
```
[the remark this answers is not in the document]
Comment (DavidLandup0) [reply keras_hub/src/models/smollm3/smollm3_utils.py]: Addressed.

File: keras_hub/src/models/smollm3/smollm3_causal_lm_preprocessor.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (DavidLandup0) [reply keras_hub/src/models/smollm3/smollm3_causal_lm_preprocessor.py:77]: Updated

File: keras_hub/src/models/smollm3/smollm3_backbone.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    model = keras_hub.models.SmolLM3Backbone.from_preset("...")
+    model(input_data)
+
+    # Randomly initialized SmolLM3 decoder with custom config.
+    model = keras_hub.models.SmolLM3Backbone(
+        ...
+    )
+    model(input_data)
+    ```
+    """
```
[the remark this answers is not in the document]
Comment (DavidLandup0) [reply keras_hub/src/models/smollm3/smollm3_backbone.py]: Updated

**Comment by DavidLandup0** [comment]
@divyashreepathihalli - looks like keras 3.8 didn't yet release RMSNormalization, but it's available in newer versions, so the tests pass for all nightly/stable versions, but not for the old one.

**Comment by DavidLandup0** [comment]
@divyashreepathihalli looks like GPT2/LLaMA hit an OOM on the internal build, but these are both unrelated to the current PR.

image

**Review by divyashreepathihalli — APPROVED** [review]
